source/FreeRTOS_TCP_IP.c

```diff
@@ -393,30 +413,48 @@
                 pxSocket->u.xTCP.usTimeout = 0U;
             }
         }
-        else
+
+        if( ( eTCPState == eCLOSED ) ||
+            ( eTCPState == eCLOSE_WAIT ) )
         {
-            if( ( eTCPState == eCLOSED ) ||
-                ( eTCPState == eCLOSE_WAIT ) )
+            /* Socket goes to status eCLOSED because of a RST.
+             * When nobody owns the socket yet, delete it. */
+            if( ( pxSocket->u.xTCP.bits.bPassQueued != pdFALSE_UNSIGNED ) ||
+                ( pxSocket->u.xTCP.bits.bPassAccept != pdFALSE_UNSIGNED ) )
             {
-                /* Socket goes to status eCLOSED because of a RST.
-                 * When nobody owns the socket yet, delete it. */
-                if( ( pxSocket->u.xTCP.bits.bPassQueued != pdFALSE_UNSIGNED ) ||
-                    ( pxSocket->u.xTCP.bits.bPassAccept != pdFALSE_UNSIGNED ) )
-                {
-                    FreeRTOS_debug_printf( ( "vTCPStateChange: Closing socket\n" ) );
+                FreeRTOS_debug_printf( ( "vTCPStateChange: Closing socket\n" ) );
 
-                    if( pxSocket->u.xTCP.bits.bReuseSocket == pdFALSE_UNSIGNED )
-                    {
-                        configASSERT( xIsCallingFromIPTask() != pdFALSE );
-                        vSocketCloseNextTime( pxSocket );
-                    }
+                if( pxSocket->u.xTCP.bits.bReuseSocket == pdFALSE_UNSIGNED )
+                {
+                    configASSERT( xIsCallingFromIPTask() != pdFALSE );
+                    vSocketCloseNextTime( pxSocket );
                 }
             }
         }
 
         /* Fill in the new state. */
         pxSocket->u.xTCP.eTCPState = eTCPState;
 
+        if( ( eTCPState == eCLOSE_WAIT ) && ( pxSocket->u.xTCP.bits.bReuseSocket == pdTRUE_UNSIGNED ) )
+        {
+            switch( xPreviousState )
+            {
+                case eSYN_FIRST:    /* 3 (server) Just created, must ACK the SYN request */
+                case eSYN_RECEIVED: /* 4 (server) waiting for a confirming connection request */
+                    FreeRTOS_printf( ( "Restoring a reuse socket port %u\n", pxSocket->usLocalPort ) );
+
+                    /* Go back into list mode. Set the TCP status to 'eCLOSED',
+                     * otherwise FreeRTOS_listen() will refuse the action. */
+                    pxSocket->u.xTCP.eTCPState = eCLOSED;
+                    ( void ) FreeRTOS_listen( ( Socket_t ) pxSocket, pxSocket->u.xTCP.usBacklog );
```

Comment:
MISRA will not like this.
`FreeRTOS_listen` -> `vTCPStateChange` -> `FreeRTOS_listen`.
How is this recursion stopped? What is to say that it won't keep calling itself?

Reply:
I was aware of it, but:

- When `vTCPStateChange()` calls `FreeRTOS_listen()`, the current TCP state is either `eSYN_FIRST` or `eSYN_RECEIVED`.
- `FreeRTOS_listen()` will call `vTCPStateChange()` while the current TCP state is `eCLOSED`. That can not lead to a new call to `listen()`.

Of course I could copy some of the listen() code into `vTCPStateChange()`, but I prefer to call the original API.